platform/admin/admin-configuration-configupdater/src/main/java/org/codice/ddf/admin/configuration/ConfigurationUpdater.java

```diff
@@ -0,0 +1,322 @@
+              () -> {
+                doHandleStore(context);
+                return null;
+              });
+    } catch (PrivilegedActionException e) {
+      // Only caught if the action threw a CHECKED exception, thus it MUST be an IOException.
+      throw (IOException) e.getCause();
+    }
+  }
+
+  /**
+   * Sync the config files in etc when an update occurs. There are four cases, which will be denoted
+   * using 2-tuple combinations of variables: [felixFileFromCache, felixFileFromConfig]
+   *
+   * <p>[null, null] (1) Config was created from config admin directly; not by dropping a file.
+   * There is nothing to track.
+   *
+   * <p>[null, X] (2) Some config was just created by dropping a file into etc; track the file.
+   * Update the properties on disk if necessary and perform any special first time processing.
+   *
+   * <p>[Y, null] [Y, X] (3) Felix prop was externally changed or removed. This is an error.
+   *
+   * <p>[Y, Y] (4) Felix file prop did not change, do nothing. This is the happy path. Properties
+   * will be updated on disk if necessary.
+   */
+  private void doHandleStore(ConfigurationContext context) throws IOException {
+    String pid = context.getServicePid();
+    File fileFromConfigAdmin = context.getConfigFile();
+
+    CachedConfigData cachedConfigData = pidDataMap.get(pid);
+    File fileFromCache = (cachedConfigData != null) ? cachedConfigData.getFelixFile() : null;
+
+    if (fileFromCache == null && fileFromConfigAdmin == null) {
+      // This config doesn't have an etc file, so we ignore this case (1)
+      LOGGER.debug("Was not tracked and will not track pid {}", pid);
+      return;
+    }
+
+    String appropriatePid =
+        (context.getFactoryPid() == null) ? context.getServicePid() : context.getFactoryPid();
+
+    if (fileFromCache == null) {
+      // An etc config file was just dropped and we're seeing it for the first time (2)
+      LOGGER.debug("Tracking pid {}", pid);
+      CachedConfigData createdConfigData = new CachedConfigData(context);
+      pidDataMap.put(pid, createdConfigData);
+      writeIfNecessary(
+          appropriatePid, fileFromConfigAdmin, context.getSanitizedProperties(), createdConfigData);
+      return;
+    }
+
+    if (!Objects.equals(fileFromConfigAdmin, fileFromCache)) {
+      if (fileFromConfigAdmin != null) {
+        // The felix file prop changed, which is not allowed (3)
+        String msg =
+            format(
+                "Felix filename has been illegally changed from [%s] to [%s]",
+                fileFromCache, fileFromConfigAdmin);
+        throw new IllegalStateException(msg);
+      }
+      if (fileFromCache.exists()) {
+        // The felix file prop was removed, which we can revert if the file exists (3)
+        LOGGER.info("Felix filename has been illegally removed, reverting to [{}]", fileFromCache);
+        context.setProperty(FELIX_FILENAME, fileFromCache.getAbsolutePath());
+        writeIfNecessary(
```

Comment:
✏️  just add a quick comment that the cache will be updated by writeIfNecessary() if necessary 😉

Reply:
if we need a comment then my method structure is bad - refactored it for you to make it more obvious what's happening